### ISSUE TITLE
#[must_use] holes

### ISSUE THREAD

**Issue by tamird** [rust-lang/rust#26281]
There are at least two problems with `#[must_use]`:
- it doesn't work on type aliases at all.
- for methods that return type containing a `#[must_use]` type, no warning is issued if the `#[must_use]` value is immediately discarded by destructuring assignment at the caller.

http://is.gd/TEG45t

**Comment by Thiez** [comment]
Performing `let _ = yields_must_use();` already silences the `#[must_use]` warning, surely it would not make sense for `let (x, _) = yields_tuple_of_u8_and_must_use();` to give a warning; it is doing the exact same thing as assigning a `#[must_use]` to `_`.

As for the type aliases, they *always* have the exact same behaviour as their original type. If you want different behaviour you must use a newtype. It would be inconsistent to make an exception for `#[must_use]`.

**Comment by tamird** [comment]
Alright, let's forget about the type alias case.

I still feel that the tuple cast (or any other destructuring assignment) presents a hole in the semantics of `#[must_use]`.

**Comment by nagisa** [comment]
I’m on the same page as @Thiez.

**Comment by Mark-Simulacrum** [comment]
The type alias problem here appears to have been discussed and decided on as intentional. The tuple problem is covered by #26291, as far as I can tell, or should be added to that issue and this issue closed.

**Comment by Mark-Simulacrum** [comment]
Just ran into `#[must_use]` on a type alias *silently* doing nothing. I understand why it doesn't make sense for it to work, but I do expect some form of warning. Should I open another issue about that?

**Comment by Mark-Simulacrum** [comment]
Rebranding this as `#[must_use]` silently does nothing on type aliases.

**Comment by zackmdavis** [comment]
> but I do expect some form of warning

Related/likely-prerequisite: #14407.

**Comment by xfix** [comment]
I don't think it would be necessarily inconsistent to make `#[must_use]` work for `type`s because unlike other attributes, `#[must_use]` doesn't affect code generation. Other attributes that don't affect code generation - `cfg`, `cfg_attr`, `allow`, `warn`, `deny`, `forbid`, `deprecated` and `doc` work fine for `type` statements.

**Comment by jieyouxu** [comment]
As of stable 1.76.0, rustc emits the following warning:

```
warning: `#[must_use]` has no effect when applied to a type alias
 --> src/main.rs:4:1
  |
4 | #[must_use]
  | ^^^^^^^^^^^
  |
  = note: `#[warn(unused_attributes)]` on by default
```
